js_modules/dagster-ui/packages/ui-core/src/asset-selection/input/useAssetSelectionAutoCompleteProvider.oss.tsx

```diff
@@ -0,0 +1,40 @@
+import {useMemo} from 'react';
+
+import {attributeToIcon, getAttributesMap} from './util';
+import {AssetGraphQueryItem} from '../../asset-graph/useAssetGraphData';
+import {createSelectionAutoComplete} from '../../selection/SelectionAutoComplete';
+import {
+  SelectionAutoCompleteProvider,
+  createProvider,
+} from '../../selection/SelectionAutoCompleteProvider';
+
+export function useAssetSelectionAutoCompleteProvider(
+  assets: AssetGraphQueryItem[],
+): SelectionAutoCompleteProvider {
+  const attributesMap = useMemo(() => getAttributesMap(assets), [assets]);
+
+  const baseProvider = useMemo(
+    () =>
+      createProvider({
+        attributesMap,
+        primaryAttributeKey: 'key',
+        attributeToIcon,
+      }),
+    [attributesMap],
+  );
+  const selectionHint = useMemo(() => createSelectionAutoComplete(baseProvider), [baseProvider]);
+
+  return {
+    ...baseProvider,
+    useAutoComplete: ({line, cursorIndex}) => {
+      const autoCompleteResults = useMemo(
+        () => selectionHint(line, cursorIndex),
+        [line, cursorIndex],
+      );
+      return {
+        autoCompleteResults,
+        loading: false,
+      };
+    },
+  };
```

Comment:
May not have an impact but it'd be nice to `useMemo` this returned object. it doesn't feel like the sort of thing that would be recreated every render, but it is because of the spread.

Reply:
We destructure the object and only use the `useAutoComplete` field.